src/numerics/Newton.cpp

```diff
@@ -0,0 +1,264 @@
+        // calculate perturbed residual
+        vector_fp xdotPerturbed(m_nv); //make this member for speed?
+        m_residfunc->eval(0, x, xdotPerturbed.data(), 0);
+
+        // compute nth column of Jacobian
+        for (size_t m = 0; m < m_nv; m++) {
+            m_jacobian.value(m,n) = (xdotPerturbed[m] - xdot[m])/dx;
+        }
+        // restore solution vector
+        x[n] = xsave;
+    }
+
+    // timestep components
+    if (m_rdt > 0) {
+        for (size_t i = 0; i < m_nv; i++) {
+            m_jacobian.value(i,i) -= m_rdt;
+        }
+    }
+
+    // for constant-valued components: 1 in diagonal position, all 0's in row and column
+    for (size_t i : m_constantComponents) {
+        for (size_t j = 0; j < m_nv; j++) {
+            m_jacobian.value(i,j) = 0;
+            m_jacobian.value(j,i) = 0;
+        }
+        m_jacobian.value(i,i) = 1;
+    }
+
+    // factor and save jacobian, will be reused for faster step computation
+    m_jacFactored = m_jacobian;
+    Cantera::factor(m_jacFactored);
+}
+
+// RMSD (weighted)
+doublereal Newton::weightedNorm(const doublereal* x, const doublereal* step) const
+{
+    double square = 0.0;
+    for (size_t i = 0; i < m_nv; i++) {
+        square += pow(step[i]/(x[i] + m_config->atol[i]), 2);
+    }
+    return sqrt(square/m_nv);
+}
+
+void Newton::step(doublereal* x, doublereal* step)
+{
+    m_residfunc->eval(0, x, step, 0);
+    for (size_t n = 0; n < m_nv; n++) {
+        step[n] = -step[n] + m_rdt*(x[n]-m_xlast[n]);
+    }
+
+    DenseMatrix solvejac = m_jacFactored;
+    try {
+        Cantera::solveFactored(solvejac, step);
+    } catch (CanteraError&) {
+        // int iok = m_jac->info() - 1;
+        int iok = -1; //TODO: enable error info
+        if (iok >= 0) {
+            throw CanteraError("Newton::step",
+                "Jacobian is singular for component {} (Matrix row {})",
+                iok, iok); //TODO: add component name
+        } else {
+            throw;
+        }
+    }
+}
+
+int Newton::hybridSolve() {
+    int MAX = 17;
+    int newtonsolves = 0;
+    int timesteps = 0;
+
+    // initial state
+    m_residfunc->getState(m_x.data());
+    copy(m_x.begin(), m_x.end(), m_xsave.begin());
+
+    for(int i = 0; i < MAX; i++) {
+        newtonsolves++;
+        m_config = &m_directsolve_config;
+        if(solve(m_x.data()) > 0) {
+            writelog("\nConverged in {} newton solves, {} timesteps.", newtonsolves, timesteps);
+            return 1;
+        }
+        m_config = &m_timestep_config;
+        copy(m_xsave.begin(), m_xsave.end(), m_x.begin());
+        for(int j = 0; j < MAX; j++) {
+            if (solve(m_x.data()) < 0) {
+                writelog("\nTimestep failure after {} newton solves, {} timesteps.", newtonsolves, timesteps);
+                return -1;
+            }
+            timesteps++;
+        }
+        copy(m_x.begin(), m_x.end(), m_xsave.begin());
+    }
+    writelog("Failure to converge in {} steps.", MAX);
+    return 0;
```

Comment:
I think this solver loop only does a much more limited version of the hybrid time-stepping / steady solver structure that's implemented by the `MultiNewton` class, and is only alternating between solving the steady state and taking time steps of a fixed size. Given the sophistication of the `MultiNewton` algorithm, it might be worth seeing if we can pose this problem in a way that that solver can use directly.